### PR TITLE
Overlap-save convolution

### PR DESCRIPTION
**Pull request by galenlynch**
I have added an overlap-save algorithm to convolution, which is substantially faster (~2x) for large convolution problems. It is also substantially faster than other packages like ImageFiltering.jl. For smaller convolutions, it is about the same or a little bit slower than standard fft convolution. The cutoff between overlap-save and simple fft convolution is pretty arbitrary at the moment, so it would be nice to add some benchmarking to allow DSP.jl to choose between the two algorithms depending on the size of the input. It would also be nice to add spatial/time domain filtering for small convolutions, as it would almost certainly be faster than fft convolutions. This is relevant to #224.

### CODE REVIEW

**Comment by galenlynch** [comment]
I think this still has bugs in it. Let me fix it up before people spend too much time on it.

File: src/dspbase.jl [hunk 1]
```diff
@@ -117,76 +117,315 @@ function deconv(b::StridedVector{T}, a::StridedVector{T}) where T
...
+            edge_blocks,
+            all_dims,
+            save_blocksize,
+            su,
+            sv,
+            nffts,
+            out_start,
+            out_stop,
+            u_start
+        )
```
Comment (martinholters) [inline src/dspbase.jl]: Could this be made a closure to avoid passing a gazillion parameters? (Of course, not doing dynamic dispatch here and just inlining that function would be even better, but I guess the type instability would be pretty bad for performance and so would be having `these_blocks` be a vector?)

Comment (galenlynch) [reply src/dspbase.jl]: Yeah, the problem is `Val{n_edges}()`, which is obviously not type stable. Using simple integers instead of `Val` types causes the code generated by `subsets` to be less than great, and more importantly the output of `subsets`, `edge_dims`, would no longer a `Tuple`, which leads to type instability. I could play around with it a bit more to see if I can out a better solution. I hadn't though of using a closure. In my previous experience with closures, it ends up forcing to compiler to do a lot of boxing and unboxing of captured variables. I'll try that.

Comment (galenlynch) [reply src/dspbase.jl]: You're also right with `these_blocks`, it's a big performance win for it to be a `Tuple`, because then the output of the subsequent `ProductIterator` is type stable.

Comment (martinholters) [reply src/dspbase.jl]: Usually variables that are only assigned to once should be safe to capture in a closure.

**Comment by galenlynch** [comment]
OS_performance

**Comment by galenlynch** [comment]
os_grid

**Comment by galenlynch** [comment]
os_grid_memory

**Comment by galenlynch** [comment]
The performance trade-off over regular fft convolution is pretty complicated.

**Comment by galenlynch** [comment]
Which I guess makes sense because how the fft size is being chosen doesn't take into account how many extra padding samples need to be transformed.

**Comment by galenlynch** [comment]
os_grid_rt

**Comment by galenlynch** [comment]
os_grid_mem

**Comment by galenlynch** [comment]
All of the cases where overlap-save performance is worse than FFT performance would be correctly handled by the heuristic in `_conv_fft!`.

**Comment by ssfrr** [comment]
What is `N` in those plots?

**Comment by galenlynch** [comment]
N is the number of dimensions (vector, matrix, 3-d array, 4-d array)

**Comment by ssfrr** [comment]
Ah, thanks. These plots are super useful! Thanks for making them. Would it be a pain to generate the same plot but color each grid space red or blue based on the heuristic in `_conv_fft!`, so we can see how the heuristic choice relates to the memory/runtime performance?

**Comment by galenlynch** [comment]
os_grid_heuristic

Blue is when the heuristic would pick overlap-save, red is when it would pick fft convolution.

**Comment by galenlynch** [comment]
It's not perfect, but it does a pretty good job.

File: src/dspbase.jl (hunk 1 above)
Comment (galenlynch) [reply src/dspbase.jl]: I tried a closure with variables being modified passed as arguments, but it still substantially changed the output of `@code_warntype`, and had more type instability for some reason.

**Comment by ssfrr** [comment]
Looks pretty good to me. That's awesome, thanks for the visualization. Is the PR ready for review?

**Comment by galenlynch** [comment]
I think so? The style is a little sloppy but it doesn't bug.

File: src/dspbase.jl (hunk 1 above)
Comment (galenlynch) [reply src/dspbase.jl]: I guess I also don't like closures with a bunch of closed-over variables from a stylistic perspective. For me, it's hard to spot the closed-over variables, hard to separate out the closure and figure out what's going on with it or use things like `@code_warntype` on closures. Basically, I'd rather have a huge parameter list than a huge set of captured variables.

File: test/dsp.jl [hunk 2]
```diff
@@ -55,6 +55,8 @@ end
         @test conv(f32a, b) ≈ fexp
         @test conv(fb, a) ≈ fexp
 
+        conv(zeros(25), zeros(4))
+
```
[the remark this answers is not in the document]
Comment (galenlynch) [reply test/dsp.jl]: Oh that's a weird edge case that I found, it turns out you can have more than one block that needs padding on each edge. I don't know how to test for something not throwing an error.

Yeah I hear you with the tests. One thing to keep in mind is that most (all?) of the existing tests for convolution end up hitting the overlap-save code, so 1d, 2d, and 3d cases are already exercised, as are single-block and multi-block, as is complex and real, etc...

File: src/dspbase.jl [hunk 3]
```diff
@@ -118,75 +118,319 @@ function deconv(b::StridedVector{T}, a::StridedVector{T}) where T
...
+function _conv_similar(u, outsize, axesu, axesv)
     out_offsets = first.(axesu) .+ first.(axesv)
-    out_axes = UnitRange.(out_offsets, out_offsets .+ minpad .- 1)
-    out = similar(y, out_axes)
-    copyto!(out, CartesianIndices(out), y, CartesianIndices(UnitRange.(1, minpad)))
+    out_axes = UnitRange.(out_offsets, out_offsets .+ outsize .- 1)
+    similar(u, out_axes)
+end
+function _conv_similar(
+    u, outsize, ::NTuple{<:Any, Base.OneTo{Int}}, ::NTuple{<:Any, Base.OneTo{Int}}
```
[the remark this answers is not in the document]
Comment (galenlynch) [reply src/dspbase.jl]: Thanks!

File: src/dspbase.jl [hunk 4]
```diff
@@ -118,75 +118,319 @@ function deconv(b::StridedVector{T}, a::StridedVector{T}) where T
...
+        copyto!(out, block_out_region, tdbuff, valid_buff_region)
+    end
+
+    out
 end
 
-function _conv(
-    u::AbstractArray{T, N}, v::AbstractArray{T, N}, paddims
+function _conv_kern_fft!(
+    out, u::AbstractArray{T, N}, v::AbstractArray{T, N}, su, sv, outsize, nffts
```
[the remark this answers is not in the document]
Comment (galenlynch) [reply src/dspbase.jl]: Having the same element type lets you use the same plan and padding buffer for both inputs.

File: src/dspbase.jl [hunk 5]
```diff
@@ -118,75 +118,319 @@ function deconv(b::StridedVector{T}, a::StridedVector{T}) where T
...
+        nfft = nextfastfft(nfull)
+    end
+
+    nfft
+end
+
+@inline function os_prepare_conv(u::AbstractArray{T, N}, nffts) where {T<:Real, N}
+    tdbuff = similar(u, nffts)
+    size_arr = collect(nffts)
+    size_arr[1] = size_arr[1] >> 1 + 1
```
[the remark this answers is not in the document]
Comment (galenlynch) [reply src/dspbase.jl]: lol my laziness about kitting out my emacs to work with unicode makes me a unicode grump by default. Great suggestion!

File: src/dspbase.jl [hunk 6]
```diff
@@ -118,75 +118,319 @@ function deconv(b::StridedVector{T}, a::StridedVector{T}) where T
...
+    fdbuff = similar(u, Complex{T}, NTuple{N, Int}(size_arr))
+
+    p = plan_rfft(tdbuff)
+    ip = plan_brfft(fdbuff, nffts[1])
+
+    tdbuff, fdbuff, p, ip
+end
+
+@inline function os_prepare_conv(u::AbstractArray{<:Complex, <:Any}, nffts)
```
[the remark this answers is not in the document]
Comment (galenlynch) [reply src/dspbase.jl]: Ah right, good catch!

File: test/dsp.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (galenlynch) [reply test/dsp.jl]: Yeah I think all of your suggested test cases are already exercised by the existing convolution tests...

However, I will think of some more explicit tests to add.

File: src/dspbase.jl [hunk 7]
```diff
@@ -118,75 +118,319 @@ function deconv(b::StridedVector{T}, a::StridedVector{T}) where T
...
 
 # padded must start at index 1, but u can have arbitrary offset
-function _zeropad!(padded::AbstractVector, u::AbstractVector)
-    datasize = length(u)
+@inline function _zeropad!(
+    padded::AbstractVector,
+    u::AbstractVector,
+    data_dest = (1,),
+    data_region = CartesianIndices(u)
+)
```
[the remark this answers is not in the document]
Comment (galenlynch) [reply src/dspbase.jl]: Just for function parameters or everywhere?

[the remark this answers is not in the document]
Comment (ssfrr) [reply src/dspbase.jl]: When you press `C-RET` in Juno it executes the current block of code, and determines what a "block" is by indentation, so anything one would want to evaluate as a block needs to be indented. The function parameter lists are the big one because often when testing things I'll tweak the code and then re-evaluate the function to update it.

File: test/dsp.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (ssfrr) [reply test/dsp.jl]: Sounds good - maybe just add a comment above that line explaining what it's testing for.

If existing `conv` tests are using large enough data to trigger the overlap-add stuff (and also small enough to do the single-block case) then I don't think we need to duplicate that (it wasn't super clear to me looking at the coverage results that everything was tested, because are some blocks that seem to be marked as "not code"). The main thing I can think of would be fencepost and off-by-one errors that might crop up when setting up the blocks.

File: src/dspbase.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (ssfrr) [reply src/dspbase.jl]: Ah, OK. Do things still work if you try to convolve an `Array{Float64}` and an `Array{Float32}`?

[the remark this answers is not in the document]
Comment (galenlynch) [reply src/dspbase.jl]: I think at least the FFT plan is type specific, but the promotion method chaining in `conv` makes everything the same type by the time it hits these kernel methods.

File: test/dsp.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (galenlynch) [reply test/dsp.jl]: Actually I was wrong, overlap-save was hit by `conv` testing only four times. I'll add some tests that does a better job of exercising overlap-save.

File: src/dspbase.jl [hunk 8]
```diff
@@ -118,75 +118,319 @@ function deconv(b::StridedVector{T}, a::StridedVector{T}) where T
 end
 
 # padded must start at index 1, but u can have arbitrary offset
-function _zeropad!(padded::AbstractVector, u::AbstractVector)
-    datasize = length(u)
+@inline function _zeropad!(
+    padded::AbstractVector,
+    u::AbstractVector,
+    data_dest = (1,),
```
[the remark this answers is not in the document]
Comment (galenlynch) [reply src/dspbase.jl]: I put a `Tuple` type bound on it, to emphasize that a `tuple` is expected. It actually needs to be a tuple not just because it comes from size, but mostly to work with `CartesianIndices`. The reason the `Vector` method also needs a `Tuple` is to have the same interface as the ND method.

File: src/dspbase.jl [hunk 9]
```diff
@@ -118,75 +118,319 @@ function deconv(b::StridedVector{T}, a::StridedVector{T}) where T
...
+# Does convolution, will not switch argument order
+function _conv!(out, u, v, su, sv, outsize)
+    # TODO: Add spatial / time domain algorithm
+    _conv_fft!(out, u, v, su, sv, outsize)
+end
+
+# Does convolution, will not switch argument order
+function _conv(u, v, su, sv)
+    outsize = su .+ sv .- 1
+    out = _conv_similar(u, v, outsize)
```
[the remark this answers is not in the document]
Comment (galenlynch) [reply src/dspbase.jl]: Maybe now that it's only used once, but it feels like an implementation detail that the caller shouldn't need to worry about.

File: test/dsp.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (galenlynch) [reply test/dsp.jl]: I don't really get having the output size an even multiple of FFT size, because the blocks move in steps of size `nffts - sv + 1`, instead of the FFT size. I added a bunch of tests. If you can think of some particular adversarial settings that you would like to test, feel free to add them! The functions that I made in `test/dsp.jl` make it pretty easy to try different inputs.

**Comment by galenlynch** [comment]
Whoops, found a bug

File: src/dspbase.jl (hunk 9 above)
[the remark this answers is not in the document]
Comment (ssfrr) [reply src/dspbase.jl]: Given that the caller is already an internal function my preference is for less abstraction so it's easier to follow the code, but I'm happy to agree to disagree on this one.

[the remark this answers is not in the document]
Comment (galenlynch) [reply src/dspbase.jl]: Maybe it's just a poorly named function. I think it feels more natural to have an interface like `_conv_similar(u, v, outsize)` because it seems similar to `similar` and makes a new array with appropriate axes to store the convolution result between `u` and `v`, but a interface like `_conv_similar(u, outsize, axes_u, axes_v)` doesn't make too much sense to me, even if that's what the next step in the process is.

**Comment by galenlynch** [comment]
@ssfrr I had to rebase and in the process I made some your comments "outdated" when I didn't really address them. It seems like the two outstanding issues are possibly adding more tests, and possibly making the edge function a closure instead of an standalone function. This is also what @martinholters wanted.

If you can think of any more test cases, I'd be happy to add them. 

As for the closure, I think that replacing function parameters with captured variables just obscures the actual complexity, and it makes it harder for the compiler to make the same quality code. I played around a little with let blocks, but couldn't make it work. To be honest I didn't try that hard to make the code work as a closure, because of my personal preference for being explicit about what is being shared between the two functions, so it may be possible with some more effort. Would you guys be more satisfied with a standalone function if I did a better job of annotating the parameters? Happy to discuss if you guys think there is a reason to use a closure.

**Comment by ssfrr** [comment]
I can go through my comment threads and see which still apply. I already marked a few as resolved. AFAICT the testing seems to be in pretty good shape at this point, but I want to do a little playing around with the PR to see if I pop anything up.

I'm wondering if the closure issue is a sign that we should be looking for a better way to refactor that logic. Maybe it's just me but I find the edge finding and nested iteration and set manipulation pretty hard to follow. Also it means that a lot of the work of figuring out _how_ to apply the block convolution is happening in the outer function, which contributes to the large number of parameters that need to be passed to the inner function. Also the block convolution process is somewhat duplicated for the edge blocks and center blocks.

Do you think it would work to simplify the outer loop to be something like:

```julia
function _conv_kern_os!(out,
                        u::AbstractArray{<:Any, N},
                        v,
                        su,
                        sv,
                        sout,
                        nffts) where N

    # Pre-allocation
    tdbuff, fdbuff, fftplan, ifftplan = os_prepare_conv(u, nffts)

    # Transform the smaller filter
    _zeropad!(tdbuff, v)
    filter_fd = os_filter_transform!(tdbuff, p)
    filter_fd .*= 1 / prod(nffts) # Normalize once for brfft

    # nffts gives us the size of the FFT we'll perform, but only part of it
    # gets saved to the output array
    save_blocksize = min.(sout, nffts .- sv .+ 1)
    nblocks = cld.(sout, save_blocksize)
    for block_idx in CartesianIndices(nblocks)
        _conv_kern_os_block!(out, u, filter_fd, tdbuff, fdbuff,
                             fftplan, ifftplan,
                             Tuple(block_idx), save_blocksize, nffts)
    end

    out
end
```

and then have `_conv_kern_os_block!` be responsible for figuring out whether it needs to zero-pad or copy? I haven't totally thought through that part but it seems like it could first zero-out the block, then find the corners of the valid data region and use `copyto!` to copy the right data in. That would avoid needing to use `Val{N}` anywhere but still be type-stable.

I have a start on what `_conv_kern_os_block!` could look like but it'll take some more fiddling to get it right. Does this seem like a worthwhile direction to you?

**Comment by galenlynch** [comment]
Yeah the logic is pretty gnarled. I was playing around with it a bit the other day to try to simplify it and I think it should be possible to do so. One thing that could be made more compact is the calculation of `edge_range`, which is the main product of all the set iteration and manipulation, and also suffers from tuple -> mutable array -> tuple design at the moment. I tried creating the tuple directly using `ntuple` but the closure really messed up type stability, and a `let` block didn't help. At the very least I could be doing more sign posting than I am to clarify the logic.

Another reason for the large set of parameters is to re-use as many things as possible, for performance reasons. As you can no doubt tell, when I was writing the `_conv_kern_os!` function the edge function was just the loop body.

The block convolution logic is definitely duplicated between the edge and center blocks. The inner loop of the edge method could be used on center blocks as is, in fact, because it is strictly more general. However I wanted to separate the two because the logic in the center blocks is dead simple, and requires no zero padding. As far as I can tell, one of the big benefits of using overlap-save vs overlap-add is that you don't need to zero-pad for the majority of the blocks. The way that zero padding works at the moment means that there is no performance savings when you call `_zeropad!` but don't end up with any padding. I think it's a big performance win to use the simpler logic and zero padding on the center blocks, and would prefer to keep the reduced set of logic.

**Comment by ssfrr** [comment]
it seems like we could combine them but keep the performance gains by having something like:

```julia
data_region = ... # figure out where to copy the data to
if axes(data_region) != axes(out)
    fill!(out, 0.0)
end

# copy the data
```

So it would only do the zero-fill if some parts of the output weren't getting filled with data. (Using `axes` here means it's only checking that the ranges are equal, which should be super cheap. Comparing the `CartesianIndices` directly currently seems to fallback to the generic `AbstractArray` implementation, which checks all the elements.)

**Comment by galenlynch** [comment]
Maybe... I'll think about how to do that. It seems like at the end of the day it's going to be a performance question. I would like to measure how much performance is lost by doing all the extra logic and putting a branch in the inner loop.

We should also keep a few things separate. 1) Closure vs standalone function 2) number of inputs to that function, either as sneaky closed-over variables or explicit parameters 3) using a separate, simplified loop for the blocks that are provably simpler to deal with or one loop capable of handling everything. If we decide to use one loop, then there's no reason to have it be in a separate function. As it is, most of the parameters being passed are constants (see my most recent commit for some more annotation).

I might have to put this on the back burner for while, I don't think I have time to do a significant overhaul at the moment.

**Comment by galenlynch** [comment]
I tried to document the logic in more detail, using the case of a 3d cube to illustrate the major steps in the nested iteration. Does it help clarify the logic for you?

**Comment by galenlynch** [comment]
Just to give some context to why I think it may be worth-while to have a simple loop for center blocks: with `a = rand(1000, 1000)` and `b = rand(3, 3)`, which is a pretty reasonable use case for e.g. images, then `conv(a, b)` breaks up the convolution into 664 edge blocks and 27225 center blocks. Of course the performance break-even point between one more general loop for every block and a specialized loop for center blocks depends on the exact performance difference between the two loops, but with two orders of magnitude more center blocks than edge blocks, small differences may add up quickly.

**Comment by HDictus** [comment]
while working on #224 I found that calling:
`conv(zeros((4, 7, 1)), zeros((3, 3, 3)))`

raises an error:

`ERROR: BoundsError: attempt to access 6×8×1 Array{Float64,3} at index [CartesianIndex(3, 3, 3)]                                                                                                                              
Stacktrace:
 [1] throw_boundserror(::Array{Float64,3}, ::Tuple{CartesianIndex{3}}) at ./abstractarray.jl:484
 [2] checkbounds at ./abstractarray.jl:449 [inlined]
 [3] macro expansion at ./multidimensional.jl:889 [inlined]
 [4] copyto!(::Array{Float64,3}, ::CartesianIndices{3,Tuple{UnitRange{Int64},UnitRange{Int64},UnitRange{Int64}}}, ::Array{Float64
,3}, ::CartesianIndices{3,Tuple{Base.OneTo{Int64},Base.OneTo{Int64},Base.OneTo{Int64}}}) at ./multidimensional.jl:884           
 [5] _zeropad! at /home/hugs/Documents/projects/misc/DSP.jl/src/dspbase.jl:157 [inlined]
 [6] _zeropad! at /home/hugs/Documents/projects/misc/DSP.jl/src/dspbase.jl:155 [inlined] (repeats 2 times)
 [7] unsafe_conv_kern_os!(::Array{Float64,3}, ::Array{Float64,3}, ::Array{Float64,3}, ::Tuple{Int64,Int64,Int64}, ::Tuple{Int64,I
nt64,Int64}, ::Tuple{Int64,Int64,Int64}, ::Tuple{Int64,Int64,Int64}) at /home/hugs/Documents/projects/misc/DSP.jl/src/dspbase.jl:
423                                                                                                                             
 [8] _conv_fft!(::Array{Float64,3}, ::Array{Float64,3}, ::Array{Float64,3}, ::Tuple{Int64,Int64,Int64}, ::Tuple{Int64,Int64,Int64
}, ::Tuple{Int64,Int64,Int64}) at /home/hugs/Documents/projects/misc/DSP.jl/src/dspbase.jl:553                                  
 [9] _conv! at /home/hugs/Documents/projects/misc/DSP.jl/src/dspbase.jl:577 [inlined]
 [10] _conv at /home/hugs/Documents/projects/misc/DSP.jl/src/dspbase.jl:584 [inlined]
 [11] conv(::Array{Float64,3}, ::Array{Float64,3}) at /home/hugs/Documents/projects/misc/DSP.jl/src/dspbase.jl:600
 [12] top-level scope at none:0`
on this branch as well as the one I'm working on.
I might figure out the cause tomorrow or later, but it wasn't immediately obvious to me in any case. Perhaps you'll figure it out before then.

**Comment by galenlynch** [comment]
The root cause was `optimalfftfiltlength` swapping inputs, which should not have been allowed. I fixed it and added a test. I also turned off `@inbounds` until we're done testing the PR.

**Comment by galenlynch** [comment]
Any progress Spencer? No rush, I understand that we're all busy.

**Comment by ssfrr** [comment]
Sorry, I'm on vacation this week. I'll take another look when I get back next week.

**Comment by galenlynch** [comment]
Rebased because master changed.

**Comment by galenlynch** [comment]
Gentle bump.

**Comment by galenlynch** [comment]
Bump.

**Comment by galenlynch** [comment]
Thanks for taking another look at this, and sorry that I've been so insistent. I'll test out the simplified logic that you've been suggesting. You're definitely right that this PR is pretty complicated, and I can see why it's hard to follow the logic that calculates the edge set.

**Comment by ssfrr** [comment]
> sorry that I've been so insistent

no need to apologize. This squeaky wheel deserved the grease. :slightly_smiling_face:

File: src/dspbase.jl [hunk 10]
```diff
@@ -170,90 +170,490 @@ function deconv(b::StridedVector{T}, a::StridedVector{T}) where T
...
+                        sout,
+                        nffts) where N
+    u_start = first.(axes(u))
+    out_axes = axes(out)
+    out_start = first.(out_axes)
+    out_stop = last.(out_axes)
+    ideal_save_blocksize = nffts .- sv .+ 1
+    # Number of samples that are "missing" if the valid portion of the
+    # convolution is smaller than the output
+    sout_deficit = max.(0, ideal_save_blocksize .- sout)
```
[the remark this answers is not in the document]
Comment (galenlynch) [reply src/dspbase.jl]: Good catch, it should read "if the output is smaller than the valid portion..."

**Comment by btmit** [comment]
Quick vote of support for this PR.

I was wondering what you folks thought the best way to apply this to a streaming application would be?

Specifically one where you only need to prepare the buffers, fft plans, etc once and then input blocks continuously without a priori knowledge of how many blocks will be received

**Comment by ssfrr** [comment]
> I was wondering what you folks thought the best way to apply this to a streaming application would be?

Yeah, that would be awesome. A lot of the necessary DSP infrastructure is here, but it would take some more design around how to represent the incoming/outgoing streams, as well as some engineering around handling the start/end conditions. I've got a lot of thoughts but it's a bit off-topic for this PR - feel free to open an issue referencing this if you want to discuss.

**Comment by ssfrr** [comment]
Hey @galenlynch, I wanted to check in on this PR. Have you had a chance to see if can be simplified?

I don't want this to languish too long and be difficult to merge in, or block other things from merging, so if you don't think you'll be able to revamp it within a week or two maybe it's best to merge it as-is (with whatever rebasing is necessary to get it on current master) and fix it up later? Functionally I think it's a strict improvement over what we have now and I don't want this great work to go to waste. I think the comments you added and the discussion in this thread should hopefully be enough to pick it up again.

Thoughts from other maintainers?

**Comment by galenlynch** [comment]
I haven't had a chance sadly, even though I agree with you that it should be possible to simplify it. Next week or two would definitely be impossible. If you're ok merging as is and then maybe simplifying it later then I agree that given the circumstances that would be the best. I'll rebase it.

**Comment by galenlynch** [comment]
@ssfrr sorry for the delay. I tried to squeeze this in before my defense, but the merge became non-trivial so I gave up. Now that I defended, I should finally have the time to get this merged!

**Comment by ssfrr** [comment]
No problem. :tada: :shield: :exclamation:

**Comment by galenlynch** [comment]
I finally got around to rebasing this. All tests are passing, but it's been a minute since I last thought about this stuff in depth, so I think I want to think it over before going ahead with the merge.

The rebase required a bit of tweaking of the the zeropad functions to allow both arbitrary indexing while also being able to specify the source and destination regions for the padded data.

I fully intend to clean up this PR by possibly eliminating the separate functions for the edges and the center of the convolution, which would also get rid of the huge number of parameters that are passed between these functions. Both @martinholters and @ssfrr have raised concerns with the readability of the code as it stands, so I think this clean-up would be quite helpful. However, as Spencer said, this PR is a big performance improvement so it might be nice to just have these changes in the code base in their current functioning if inelegant form, and I can work on cleaning it up in a separate PR.

**Comment by ssfrr** [comment]
Awesome, glad you're back on the case. Regarding the arbitrary indexing stuff - I wonder if there's an approach where we use regular Arrays internally and then re-apply the offset at the end of the process? It seems like maintaining all the offset stuff creates a lot of complexity but isn't actually relevant for the convolution itself. Just a thought.

**Comment by galenlynch** [comment]
That's a possibility, but I think it would just move the complexity around. Right now, only `_zeropad` methods have to be aware of arbitrary indexing, and do so with a single method for both 1-indexed arras as well as arrays with arbitrary offsets. It just looks at the axes of every array to determine the first index, instead of assuming it's one. The `conv` methods are otherwise pretty agnostic about indexing. We could add another method that copies the data, but that would be more memory intensive and possibly slower. Also, having `zeropad` and similar functions handle arbitrary indexing is also the approach that we took in #279, so we would have to change those functions as well if we were to approach the problem by copying results into the desired array format at the end.

**Comment by galenlynch** [comment]
OK I think it's ready to merge if you're happy with it.

File: src/dspbase.jl [hunk 11]
```diff
@@ -117,90 +117,490 @@ function deconv(b::StridedVector{T}, a::StridedVector{T}) where T
...
+    while pow2 <= prev_pow2
+        new_complexity = os_fft_complexity(2 ^ pow2, nb)
+        new_complexity > prev_complexity && break
+        prev_complexity = new_complexity
+        pow2 += 1
+    end
+    nfft = pow2 > prev_pow2 ? 2 ^ prev_pow2 : 2 ^ (pow2 - 1)
+
+    # L is the number of usable samples produced by each block
+    L = nfft - nb + 1
```
[the remark this answers is not in the document]
Comment (galenlynch) [reply src/dspbase.jl]: You're right! Nice catch. I made a commit that implements your suggestion.

**Comment by galenlynch** [comment]
@ssfrr good to merge?

File: src/dspbase.jl (hunk 10 above)
[the remark this answers is not in the document]
Comment (galenlynch) [reply src/dspbase.jl]: Done

**Comment by ssfrr** [comment]
:tada: Thanks for putting up with my sluggish response times and sticking with this!

**Comment by galenlynch** [comment]
Thanks for putting in all the time to read my confusing code! The end result is much better thanks to your comments and holding my feet to the fire.
